src/config.rs

```diff
@@ -0,0 +1,40 @@
+use crate::state::State;
+use serde::Deserialize;
+use sqlx::postgres::PgPool;
+use std::sync::Arc;
+
+#[derive(Deserialize, Clone)]
+pub struct PostgresConfig {
+    host: String,
+    user: String,
+    password: String,
+    db: String,
+}
+
+impl PostgresConfig {
+    pub fn database_url(&self) -> String {
+        format!(
+            "postgres://{}:{}@{}/{}",
+            self.user, self.password, self.host, self.db
+        )
+    }
+}
+
+#[derive(Deserialize, Clone)]
+pub struct Config {
+    pub postgres: PostgresConfig,
+}
+
+impl Config {
+    pub fn from_env() -> anyhow::Result<Self> {
+        let mut cfg = config::Config::new();
+        cfg.merge(config::Environment::new().separator("_"))?;
+        Ok(cfg.try_into()?)
+    }
+
+    pub async fn into_app_state(self) -> anyhow::Result<Arc<State>> {
+        let db = PgPool::connect(&self.postgres.database_url()).await?;
+
+        Ok(Arc::new(State { db }))
+    }
```

Comment:
Rust brings `traits` to achieve a conversion like this, thus I suggest implementing the `TryFrom` `trait` for `State`, to be created from a `Config` instance.

Another consideration is the return type for this method.

Even when we need to use an atomically reference counted because of the database pool reference, returning it straight from the constructor doesn't looks right, why if in the future we don't want to use an `Arc` anymore? We would have to change this method as well. I think we should return the `State` instance to the caller and let the caller decide how ownership and memory handling is done.

Reply:
In this case, I am not sure that we can use that trait due to the async nature of this convertion. What do you think?

Reply:
You can do it two steps. Here two important task are being mixed in one function.

1. The `Config` creation and validation
2. The Database connection (Pool)

What do you think about doing:

1. Create a `Config` instance (or `State` which is more accurate as `Actix` expects state to be shared not configuration which must be `static`. By implementing a function that reads config input, environment variables for instance, and constructs a `State` instance with a `PgPool` reference which is not `connected` yet.

2. The same `State` could have a method to initialize itself, where we make sure everything is settled and perform the connection to the database.

So, if `State::new` fails, then its because some value provided to the `State` is not the expected, say a environment variable is missing .

And for the case where `State::init` fails, its probably because there's some trouble with some resource like the database connection for instance.

Reply:
I created a new type:

pub type AppState = std::sync::Arc<State>;

Reply:
I didn't see your message when I wrote the last message, I will try that using the Builder Pattern!

Reply:
I don't think this is a use case for the builder pattern. The builder patter is used for complex objects which require a configuration like `clap::App` for instance.

It looks like we are _drowning in a glass of water_, we don't need to know the database configuration all around the application, we just need to be able to communicate with the database, that means consuming database connection via reference, that should be just enough